### PR TITLE
Added options to add systemd unit and create users

### PR DESCRIPTION
**Pull request by astorath**
changes:
- feat: option to add systemd service unit
- feat: option to add users on package install
- feat: options to define files and directory ownership in config

#55 #15

### CODE REVIEW

**Comment by caarlos0** [comment]
>     * feat: option to add systemd service unit
> 
>     * feat: option to add users on package install
> 
>     * feat: options to define files and directory users in config

hey, this PR has at least 3 features it seems, can we split them in one PR each?

**Comment by astorath** [comment]
> hey, this PR has at least 3 features it seems, can we split them in one PR each?

@caarlos0 problem is, 
1. they are non-independent
2. they all use the same code for scripting

**Comment by caarlos0** [comment]
you're right.

thinking here: aren't all those things here already possible with the current nfpm version?

is just custom scripts, right? If so, I don't think we should add this to nfpm itself...

**Comment by astorath** [comment]
> thinking here: aren't all those things here already possible with the current nfpm version?
> 
> is just custom scripts, right? If so, I don't think we should add this to nfpm itself...

@caarlos0 

well, consider this:

1. You can't manage files ownership in current version
2. Many (if not most) users don't even know how these scripts look like: rpm as well as deb have a set of helper units providing these functions: https://github.com/systemd/systemd/blob/master/src/core/macros.systemd.in, https://packages.debian.org/jessie/dh-systemd, so you never have to write those scripts by hand
3. Scripts are basically the same all the time, no need change anything
4. It's an fpm option already

**Comment by astorath** [comment]
Added references to issues to track for relevance

**Comment by caarlos0** [comment]
> * You can't manage files ownership in current version

this I agree we should add

> * Many (if not most) users don't even know how these scripts look like: rpm as well as deb have a set of helper units providing these functions: https://github.com/systemd/systemd/blob/master/src/core/macros.systemd.in, https://packages.debian.org/jessie/dh-systemd, so you never have to write those scripts by hand

if we add systemd, it would make sense to add all the other possible init systems as well, which I rather not do, so that's why maybe its better we don't add any...

> * Scripts are basically the same all the time, no need to change anything

until someone wants to do something different for some reason...

> * It's an fpm option already

this is not fpm, as the name says hehe

**Comment by caarlos0** [comment]
cc/ @djgilcrease any thoughts on this?

File: deb/deb.go [hunk 1]
```diff
@@ -389,3 +486,19 @@ func writeControl(w io.Writer, data controlData) error {
 	})
 	return template.Must(tmpl.Parse(controlTemplate)).Execute(w, data)
 }
+
+func getFilesAttr(raw string) (name, user, mode string) {
```
Comment (djgilcrease) [inline deb/deb.go]: this should move somewhere shareable by both rpm and deb

```go
func GetFilesAttr(raw string) (name, user, mode string) {
	var parts []string
        parts = append(parts, strings.SplitN(raw, ":", 3)...)
        parts = append(parts, "", "")
        return parts[0], parts[1], parts[2]
}
```

**Comment by djgilcrease** [comment]
I do not overly mine only supporting systemd, but I see a few issues, if I specify a systemd unit file and already have scripts they would not be merged.

I would prefer a sub command to generate systemd units and associated scripts as actual files so the user can just add them explicitly to the config file or merge it wither their own scripts

**Comment by caarlos0** [comment]
> I would prefer a sub command to generate systemd units and associated scripts as actual files so the user can just add them explicitly to the config file or merge it wither their own scripts

this makes more sense to me as well.

**Comment by astorath** [comment]
@caarlos0  @djgilcrease 
Ok, guys, before I move on to fixing issues, tests etc... Let's fix up what we are doing. To sum up:

- `systemd` is the only service manager on supported versions of RHEL / Amazon Linux 2 /  Debian / Ubuntu, so I don't see any need to support something else (`init`)
- installing program as a linux service is very much vital part of many packages
- `fpm` as well as `rpmbuild` and `dpkg` provides an easy way to do this without much effort or diving into specifics
- adding `systemd` service is not only about hook scripts, but also about adding dependencies, placing service files, etc. - so according to the DRY principle this must be a separate option, not a combination of `bash` stuff on top of CI procedures...

>  if I specify a systemd unit file and already have scripts they would not be merged.
- @djgilcrease actually they would, there are even some tests for the case

**Comment by ti-mo** [comment]
Hi @astorath!

Could I suggest defaulting packaged files to be owned by `root` (at least for RPM)? I've run into some warnings while installing RPMs built by current master, see https://github.com/goreleaser/nfpm/issues/55#issuecomment-574551328 for more details.

Basically, leaving `rpmpack.RPMFile.Owner` and `.Group` empty causes packaged files to have no owner/group:
```
[root@66f5aed57836 ~]# rpm -qlvp conntracct_0.2.0-next_linux_amd64.rpm
-rw-r--r--    1                                    344 Jan 14 12:59 /etc/conntracct/conntracct.yml
-rwxr-xr-x    1                               14538208 Jan 14 13:16 /usr/bin/conntracct
```

as opposed to:
```
[root@689f5577f54f app]# rpm -qlvp conntracct_0.2.0-next_linux_amd64.rpm
-rw-r--r--    1 root     root                      344 Jan 14 12:59 /etc/conntracct/conntracct.yml
-rw-r--r--    1 root     root                      404 Jan 15 07:52 /lib/systemd/system/conntracct.service
-rwxr-xr-x    1 root     root                 14538208 Jan 15 08:15 /usr/bin/conntracct
```

This causes `rpm` to throw warnings during package install:
```
[root@66f5aed57836 app]# rpm -Uvh conntracct_0.2.0-next_linux_amd64.rpm
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
Updating / installing...
   1:conntracct-0.2.0-next            
warning: user  does not exist - using root
warning: group  does not exist - using root
################################# [100%]
```

Patch looks good so far! Adding `info.{User,Group}` also allows user/group support to be implemented in `goreleaser`. :+1:

**Comment by astorath** [comment]
> Could I suggest defaulting packaged files to be owned by root (at least for RPM)? I've run into some warnings while installing RPMs built by current master, see #55 (comment) for more details.

@ti-mo this is not a problem, but I would like see some feedback from team members. @caarlos0 ?

**Comment by caarlos0** [comment]
I think that for now defaulting to root should be good enough yes 🤔

**Comment by astorath** [comment]
> I think that for now defaulting to root should be good enough yes 🤔

Sorry, maybe I should've put it more strait: I need some feedback on all this PR. See my comment on 5 dec.

**Comment by wg** [comment]
Thank you for this contribution @astorath! NFPM is great, and I need support for multiple systemd units which I've implemented here: https://github.com/kentik/nfpm. Once this PR is merged I'll be happy to submit another with my changes.

**Comment by erikgeiser** [comment]
While I appreciate the work and sentiment behind this proposal, I think it is a bad idea in the long run. I think a packaging program should be agnostic about what it is packaging and `systemd` unit files concern the subject that is being packaged rather than the process of packaging. Right now, `nfpm` does not really make assumptions about the target system for which a package is created. Interfacing with programs that may or may not be on the target system can have numerous problems: The interface or API could change, typical workflows could change, the folder structure could change, `systemd` itself could be retired or replaced (probably not gonna happen but who knows). As soon as this starts to happen `nfpm` will either drop support of some target configurations or it will need additional code paths and bloat to accommodate for these changes. At some point `nfpm` would look like `fpm` which is probably undesirable.

Also, in my opinion, magic functionality via post/pre-install scripts is way worse than just embracing new `systemd` features such as [dynamic users](http://0pointer.net/blog/dynamic-users-with-systemd.html) which would be more modern appoach. I would encourage everyone to take a look at all the [powerful features systemd provides](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) and handcraft a unit file that is tailor-made for the program you are packaging an then adding it to the files (or config files) section.

Please correct me if you feel different.

**Comment by caarlos0** [comment]
I think @erikgeiser put it in words better than I could.

That all being said, I think its time to close this.

Thanks everyone! I'm sorry this wasn't merged, but I think its for the best in the long run.

**Comment by caarlos0** [comment]
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request closed without merging**